src/object/cli_obj.c

```diff
@@ -617,16 +617,162 @@ obj_shard_find_replica(struct dc_object *obj, unsigned int target,
 }
 
 static int
-obj_grp_leader_get(struct dc_object *obj, int idx, unsigned int map_ver, uint8_t *bit_map)
+obj_ec_leader_select(struct dc_object *obj, int grp_idx, uint8_t *bit_map)
+{
+	struct daos_oclass_attr *oca;
+	struct pl_obj_shard	*pl_shard;
+	int			tgt_idx;
+	int			grp_size;
+	int			grp_start;
+	int			i;
+	int			shard;
+
+	oca = daos_oclass_attr_find(obj->cob_md.omd_id, NULL);
+	grp_size = obj_ec_tgt_nr(oca);
+	grp_start = grp_idx * grp_size;
+
+	/* 1. Find one from parity, and start from the last parity. */
+	tgt_idx = grp_size - 1;
+	for (i = 0; i < obj_ec_parity_tgt_nr(oca);
+	     i++, tgt_idx = (tgt_idx - 1 + grp_size) % grp_size) {
+
+		shard = grp_start + tgt_idx;
+		pl_shard = obj_get_shard(obj, shard);
+		if (pl_shard->po_target == -1 || pl_shard->po_shard == -1 ||
+		    pl_shard->po_rebuilding) {
+			/* Then try former one */
+			continue;
+		}
+		goto found;
+	}
+
+	/* If we do not know which data shards participate in the transaction,
+	 * let's return DER_STALE, then object I/O might refresh the pool map
+	 * and layout until parity rebuilt finish.
+	 */
+	if (bit_map == NIL_BITMAP) {
+		D_WARN(DF_OID" all parity shards %d are in rebuilding, retry later.\n",
+		       DP_OID(obj->cob_md.omd_id), obj_ec_parity_tgt_nr(oca));
+		return -DER_STALE;
+	}
+
+	/* Find one from the bitmap */
+	shard = -1;
+	tgt_idx = 0;
+	/* Choose one from the data shards within bit_map */
+	for (i = 0; i < obj_ec_data_tgt_nr(oca); i++, tgt_idx = (tgt_idx + 1) % grp_size) {
+		int shard_idx;
+
+		if (bit_map != NIL_BITMAP && isclr(bit_map, tgt_idx))
+			continue;
+
+		shard_idx = grp_start + tgt_idx;
```

Comment:
it will use shard == -1 check below, note: this loop will also be used to check if any data shards are failed.

Reply:
For the efficiency perspective, the caller itself needs to check that when prepare the bitmap before calling obj_ec_leader_select(), that will avoid double scan the object layout. On the other hand, if some parity node is available for be leader, then we will skip the check here. So I would suggest the make the caller to handle that.

Reply:
We need scan the whole data shards to make sure there are no further failure data shards before choose any data shard to be the leader.

If there are parity shard available, the data shards scan will be skipped.

or I miss your comments?

Reply:
> If there are parity shard available, the data shards scan will be skipped.

Can we guarantee that there are no data shards corrupted under such case (parity available)? Seems not, but we did not check, right? Then why we need to check that all data shards are not corrupted if parity unavailable?

Reply:
I think here it is trying to be cautious here,  since 2 parities are already failed, so it tries to do further check, since there is no point to go further if any other data shard failed.

Anyway, this is the original logic. But I think you got a point as well. I will update it in my next patch, which might cleanup the EC parity/leader selection further. I may update it in that patch. Thanks.